mlflow/metrics/genai/genai_metric.py

```diff
@@ -558,4 +604,104 @@ def eval_fn(
+        raise MlflowException(
+            message="The artifact JSON file appears to be corrupted and cannot be deserialized. "
+            "Please regenerate the custom metrics and rerun the evaluation. "
+            "Ensure that the file is correctly formatted and not tampered with.",
+            error_code=PUBLIC_UNDOCUMENTED,
+        )
+
+    if mlflow_version_at_ser != VERSION:
+        warnings.warn(
+            f"The custom metric definitions were serialized using MLflow {mlflow_version_at_ser}. "
+            f"Deserializing them with the current version {VERSION} might cause mismatches. "
+            "Please ensure compatibility or consider regenerating the metrics "
+            "using the current version.",
+            UserWarning,
+            stacklevel=2,
+        )
+
+    args_dict.pop("mlflow_version")
+    args_dict.pop("fn_name")
+    if fn_name == make_genai_metric_from_prompt.__name__:
+        return make_genai_metric_from_prompt(**args_dict)
+
+    examples = args_dict["examples"]
+    if examples is None:
+        return make_genai_metric(**args_dict)
+
+    deser_examples = []
+    for example in examples:
+        deser_examples.append(EvaluationExample(**example))
+
+    args_dict["examples"] = deser_examples
+    return make_genai_metric(**args_dict)
+
+
+def search_custom_metrics(
+    run_id: str,
+    name: Optional[str] = None,
+    version: Optional[str] = None,
+) -> List[EvaluationMetric]:
+    """
+    Searches for custom metrics created by users through `make_genai_metric()` or
+    `make_genai_metric_from_prompt()` that are associated with a particular evaluation run.
+
+    Args:
+        run_id: The unique identifier for the run.
+        name: (Optional) The name of the custom metric to search for.
+            If None, search for all metrics.
+        version: (Optional) The version of the custom metric to search for.
+            If None, search for all versions.
+
+    Returns:
+        A list of EvaluationMetric objects that match the search criteria.
+
+    .. code-block:: python
+        :test:
+        :caption: Example for searching for a custom genai metric
+
+        from mlflow.metrics.genai import search_custom_metrics
+
+        metrics = search_custom_metrics(
+            run_id=run.info.run_id, name="answer_similarity", version="v1"
+        )
+    """
+    client = mlflow.MlflowClient()
+    artifacts = [a.path for a in client.list_artifacts(run_id)]
+    if _GENAI_CUSTOM_METRICS_FILE_NAME not in artifacts:
+        _logger.warning("No custom metric definitions were found for this evaluation run.")
+        return []
+
+    with TemporaryDirectory() as tmpdir:
+        downloaded_artifact_path = mlflow.artifacts.download_artifacts(
+            run_id=run_id,
+            artifact_path=_GENAI_CUSTOM_METRICS_FILE_NAME,
+            dst_path=tmpdir,
+        )
+        custom_metrics = client._read_from_file(downloaded_artifact_path)
```

Comment:
Can we use `json.loads` instead of pandas? To avoids weirdness of Pandas type handling e.g. null, complex objects.

Reply:
oh I was reading around the codebase and thought we prefer to use pandas for read/writes. cuz the _read_from_file helper calls pd.read_json anyway

Reply:
also the log_table call actually uses pandas under the hood so thought to use pandas for read as well

Reply:
The reason those methods use pandas is that the target artifacts is table format, which is the natural data representation in Pandas (column oriented structure). Here the metric definition is more like key-value store and we don't need column oriented structure. Using Pandas to loading JSON into dictionary doesn't have any benefit.

Also sharing an example of pandas weirdness - None in float column is implicitly converted to one of `np.nan` or `pd.NaN`, and which is not same object as Python's `None`🤯

Reply:
yah I used log_table that's why I meant by using pandas to write. I agree this json file don't necessarily need pandas for read and write. The reason I used log_table was because I wanted to reduce the UI work by reusing the UI rendering view we've built for eval_results_table.json. hence I didn't add a new field but logged this artifact as a table

Reply:
as for the pandas none/nan/na stuff, @BenWilson2 made a similar comment in the previous PR [2/x]. the only field that might be None here is "version" when the metric is created from make_genai_metric_from_prompt. I tested the case where the entire column of "version" is None and the inferred the type of that pandas column is indeed a bit funny. Hence I changed to pass in empty string "" instead of None for version: https://github.com/mlflow/mlflow/blob/f416a3b685bc8fe0bfad85d3f41cb09ff8edcef2/mlflow/models/evaluation/default_evaluator.py#L1839  to enforce the column type to be object.

this test case here is for this case
https://github.com/mlflow/mlflow/blob/f416a3b685bc8fe0bfad85d3f41cb09ff8edcef2/tests/evaluate/test_default_evaluator.py#L4264-L4266

Reply:
Understood, thank you for the validation!